### ISSUE TITLE
separate run script in environment-setting and M2-binary call:

### ISSUE THREAD

**Issue by jakobkroeker** [Macaulay2/M2#247]
prurpose: set environment without starting M2 (e.g. to check linked libraries with 'ldd')

possible solution:
file 'M2':
```
exec /run/media/jakob/TOEXT4/work/home_fc18/Projects/M2-master/M2/BUILD/dan/builds.tmp/fedora32/usr-dist/i686-Linux-Fedora-18/bin/M2.run "$@"
```
file 'M2.sh'
```
cd /run/media/jakob/TOEXT4/work/home_fc18/Projects/M2-master/M2/BUILD/dan/builds.tmp/fedora32/usr-dist/i686-Linux-Fedora-18/bin/ 
. M2.sh `pwd`
exec bash
```
file 'usr-dist/i686-Linux-Fedora-18/bin/M2.run':
```
#! /bin/sh
# We assume that @ pre_bindir @/.. is ${pre_exec_prefix} (see configure.ac)
. `dirname "$0"`/M2.sh `dirname "$0"`
exec `dirname "$0"`/M2-binary "$@"
# Local Variables:
# compile-command: "make -C $M2BUILDDIR/Macaulay2/bin M2 "
# End:
```
file  'usr-dist/i686-Linux-Fedora-18/bin/M2.sh':
```
#! /bin/sh
# We assume that @ pre_bindir @/.. is ${pre_exec_prefix} (see configure.ac)
case "linux-gnu" in
   cygwin) export PATH=$1/../lib/Macaulay2/i686-Linux-Fedora-18/lib:$PATH ;;
  darwin*) # Mac OS X:
           export DYLD_LIBRARY_PATH=$1/../lib/Macaulay2/i686-Linux-Fedora-18/lib:$DYLD_LIBRARY_PATH ;;
        *) export   LD_LIBRARY_PATH=$1/../lib/Macaulay2/i686-Linux-Fedora-18/lib:$LD_LIBRARY_PATH ;;
esac
```

**Comment by DanGrayson** [comment]
Do you have a proposal?  I'm not sure what it is.

**Comment by jakobkroeker** [comment]
I would like to have a possibility to set the environment variables (LD_LIBRARY_PATH) without doing this by hand. Therefore I separated the M2 start script into parts, as above.

**Comment by DanGrayson** [comment]
This seems useful only to people debugging Macaulay2, not to people using it in the ordinary way, so I'm closing it.  Meanwhile, we might like to get rid of the use of the script entirely so M2 itself could be a script interpreter, as was the case before version 1.7.  See https://github.com/Macaulay2/M2/issues/234 .

**Comment by jakobkroeker** [comment]
>Do you have a proposal? 

https://github.com/Macaulay2/M2/pull/250

**Comment by jakobkroeker** [comment]
on the other side, it is quite useful to have this `M2_setenv` script until #234 is not done

**Comment by DanGrayson** [comment]
Suppose you make a symbolic link to M2 from some other directory.  Can you run it still?

**Comment by DanGrayson** [comment]
The files M2 and M2.sh contain absolute paths, so we can't move Macaulay2 without editing those files.

**Comment by DanGrayson** [comment]
Reopening, because you seem to have already done the work.

**Comment by jakobkroeker** [comment]
my pull request #250 needs some more work; I'm not happy with it as is..

>Reopening, because you seem to have already done the work.

Probably I should have looked at #234 and worked on that instead...

**Comment by DanGrayson** [comment]
No, I'm not sure we have a good solution for #234 that we would like to
implement.
Not every ubuntu system has chrpath installed, and not every Macaulay2 user
has root permission.

**Comment by DanGrayson** [comment]
Can you fix M2 and M2.sh so they contain no absolute paths, and so that a symbolic link to M2 from another directory will work just as well?

**Comment by jakobkroeker** [comment]
> that a symbolic link to M2 from another directory will work just as well

does this work for recent Macaulay2?

a fragile approach using `readlink` may work for symbolic links, too:
https://github.com/Macaulay2/M2/compare/master...jakobkroeker:separating.adjustEnv.and.runM2
Pleas take a look and comment. Thanks!

Caveat:
the design is overcompicated by trying to avoid duplicate code:
the various additional `M2_setenv_tail` `M2_common_setenv_bin` `M2_tail` files 
are there only for that purpose.

maybe you have a better suggestion how to avoid code duplication?

**Comment by DanGrayson** [comment]
Here's a better way to debug the setting of environment variables by the script "M2".  Change the last line from 
```
exec `dirname "$0"`/M2@EXE@ "$@"
```
to
```
M2_BINARY=`dirname "$0"`/M2@EXE@
if test "$M2_SHELL" = yes
then $SHELL
else exec "$M2_BINARY" "$@"
fi
```
Then, to debug, run
```
  M2_SHELL=yes M2
```
and issue the commands of your choosing in the shell that starts.
